### DISCUSSION TITLE
Trivy Filesystem - Number of language-specific files: 0

### DISCUSSION THREAD

**Discussion by eliassal** [aquasecurity/trivy#5808] [Q&A] [answered]
### Description

I have recently came across trivy and started to learn about it so I installed it on a ubuntu VM. I have a DOTNET webapp, when I run

```
trivy fs . > trivy-fs_report.txt 
trivy fs /home/jenkins/workspace/DevSecOps-DotNet-CI-CD-ToRunOnUbuntuNode/ > trivy-fs_report.txt 
trivy fs /home/jenkins/workspace/DevSecOps-DotNet-CI-CD-ToRunOnUbuntuNode/src > trivy-fs_report.txt 
```

I get 

> 2023-12-19T16:53:21.999+0100	INFO	Need to update DB
> 2023-12-19T16:53:21.999+0100	INFO	DB Repository: ghcr.io/aquasecurity/trivy-db
> 2023-12-19T16:53:22.000+0100	INFO	Downloading DB...
> 41.95 MiB / 41.95 MiB [--------------------------------------------------] 100.00% 4.41 MiB p/s 9.7s
> 2023-12-19T16:53:36.679+0100	INFO	Vulnerability scanning is enabled
> 2023-12-19T16:53:36.679+0100	INFO	Secret scanning is enabled
> 2023-12-19T16:53:36.680+0100	INFO	If your scanning is slow, please try '--scanners vuln' to disable secret scanning
> 2023-12-19T16:53:36.680+0100	INFO	Please see also https://aquasecurity.github.io/trivy/v0.48/docs/scanner/secret/#recommendation for faster secret detection
> 2023-12-19T16:53:47.228+0100	INFO	Number of language-specific files: 0

So should I understand that trivy does not take into consideration DOTNET App?

Here is a snapshot of the directory

![image](https://github.com/aquasecurity/trivy/assets/8447864/4671e273-401c-4eca-a281-5942bcb005fb)

### Desired Behavior

Get Vulnerabilities report on Dotnet app

### Actual Behavior

no report
trivy fs . > trivy-fs_report.txt 
trivy fs /home/jenkins/workspace/DevSecOps-DotNet-CI-CD-ToRunOnUbuntuNode/ > trivy-fs_report.txt 
trivy fs /home/jenkins/workspace/DevSecOps-DotNet-CI-CD-ToRunOnUbuntuNode/src > trivy-fs_report.txt 

### Reproduction Steps

```bash
1.trivy fs . > trivy-fs_report.txt 
2.trivy fs /home/jenkins/workspace/DevSecOps-DotNet-CI-CD-ToRunOnUbuntuNode/ > trivy-fs_report.txt
3.trivy fs /home/jenkins/workspace/DevSecOps-DotNet-CI-CD-ToRunOnUbuntuNode/src > trivy-fs_report.txt 
...
```

### Target

Filesystem

### Scanner

Vulnerability

### Output Format

Table

### Mode

Standalone

### Debug Output

```bash
trivy fs /home/jenkins/workspace/DevSecOps-DotNet-CI-CD-ToRunOnUbuntuNode/ --debug

2023-12-19T17:08:25.546+0100	DEBUG	Severities: ["UNKNOWN" "LOW" "MEDIUM" "HIGH" "CRITICAL"]
2023-12-19T17:08:25.547+0100	DEBUG	Ignore statuses	{"statuses": null}
2023-12-19T17:08:25.636+0100	DEBUG	cache dir:  /home/salam/.cache/trivy
2023-12-19T17:08:25.636+0100	DEBUG	DB update was skipped because the local DB is the latest
2023-12-19T17:08:25.636+0100	DEBUG	DB Schema: 2, UpdatedAt: 2023-12-19 12:13:54.038117392 +0000 UTC, NextUpdate: 2023-12-19 18:13:54.038117111 +0000 UTC, DownloadedAt: 2023-12-19 15:53:32.37625009 +0000 UTC
2023-12-19T17:08:25.636+0100	INFO	Vulnerability scanning is enabled
2023-12-19T17:08:25.636+0100	DEBUG	Vulnerability type:  [os library]
2023-12-19T17:08:25.636+0100	INFO	Secret scanning is enabled
2023-12-19T17:08:25.636+0100	INFO	If your scanning is slow, please try '--scanners vuln' to disable secret scanning
2023-12-19T17:08:25.636+0100	INFO	Please see also https://aquasecurity.github.io/trivy/v0.48/docs/scanner/secret/#recommendation for faster secret detection
2023-12-19T17:08:25.636+0100	DEBUG	Enabling misconfiguration scanners: [azure-arm cloudformation dockerfile helm kubernetes terraform terraformplan]
2023-12-19T17:08:25.636+0100	DEBUG	No secret config detected: trivy-secret.yaml
2023-12-19T17:08:25.636+0100	DEBUG	The nuget packages directory couldn't be found. License search disabled
2023-12-19T17:08:25.637+0100	DEBUG	Walk the file tree rooted at '/home/jenkins/workspace/DevSecOps-DotNet-CI-CD-ToRunOnUbuntuNode' in parallel
2023-12-19T17:08:26.118+0100	DEBUG	OS is not detected.
2023-12-19T17:08:26.119+0100	DEBUG	Detected OS: unknown
2023-12-19T17:08:26.119+0100	INFO	Number of language-specific files: 0
```

### Operating System

Ubuntu 22

### Version

```bash
Version: 0.48.0
Vulnerability DB:
  Version: 2
  UpdatedAt: 2023-12-19 12:13:54.038117392 +0000 UTC
  NextUpdate: 2023-12-19 18:13:54.038117111 +0000 UTC
  DownloadedAt: 2023-12-19 15:53:32.37625009 +0000 UTC
Policy Bundle:
  Digest: sha256:8bfc31f3e4301ef758b6793a07e0b12b4306e0b54d03a640efb2ff5e5ef29b9e
  DownloadedAt: 2023-12-14 16:04:48.395972148 +0000 UTC
```

### Checklist

- [ ] Run `trivy image --reset`
- [ ] Read [the troubleshooting](https://aquasecurity.github.io/trivy/latest/docs/references/troubleshooting/)

**Comment by DmitriyLewen** [comment]
Hello @eliassal 
Thanks for your report!

You can check list of supported .NET files [here](https://aquasecurity.github.io/trivy/v0.48/docs/coverage/language/dotnet/).
If your app has one of this file - please send this file and we will investigation why Trivy is not finding packages from this file.

Regards, Dmitriy

**Comment by eliassal** [comment]
@DmitriyLewen OK, I checked , yes no file was there according to your list so I tried another dotnet app which has nuget packages Now when I run trivy on this I get nothing in spite of the fact that it says "Number of language-specific files: 1"

> k:~$ trivy fs /home/salam/Projects/mslearn/
> 2023-12-20T18:47:31.492+0100	INFO	Need to update DB
> 2023-12-20T18:47:31.492+0100	INFO	DB Repository: ghcr.io/aquasecurity/trivy-db
> 2023-12-20T18:47:31.492+0100	INFO	Downloading DB...
> 42.02 MiB / 42.02 MiB [--------------------------------------------------] 100.00% 5.61 MiB p/s 7.7s
> 2023-12-20T18:47:40.665+0100	INFO	Vulnerability scanning is enabled
> 2023-12-20T18:47:40.665+0100	INFO	Secret scanning is enabled
> 2023-12-20T18:47:40.665+0100	INFO	If your scanning is slow, please try '--scanners vuln' to disable secret scanning
> 2023-12-20T18:47:40.665+0100	INFO	Please see also https://aquasecurity.github.io/trivy/v0.48/docs/scanner/secret/#recommendation for faster secret detection
> 2023-12-20T18:47:46.503+0100	INFO	To collect the license information of packages in "package-lock.json", "npm install" needs to be performed beforehand
> 2023-12-20T18:48:03.985+0100	INFO	Number of language-specific files: 1

You can download the project at "mslearn.zip" from my personal site at https://transfert.free.fr/0u1Byif
I have already npm installed, why it says ""npm install" needs to be performed beforehand"

:~$ npm -v
9.8.1

**Reply by DmitriyLewen** [reply, marked as answer]
Looks like all your dependencies are dev dependencies.
We exclude dev deps from report - https://aquasecurity.github.io/trivy/v0.48/docs/coverage/language/nodejs/#npm
Use `--include-dev-deps` to show them:
```bash
➜ trivy -d fs --include-dev-deps ./package-lock.json 
...
package-lock.json (npm)

Total: 42 (UNKNOWN: 0, LOW: 1, MEDIUM: 14, HIGH: 22, CRITICAL: 5)
...
```

> I have already npm installed, why it says ""npm install" needs to be performed beforehand"

Trivy takes liceneses from `package.json` files from `node-modules` folder - https://aquasecurity.github.io/trivy/v0.48/docs/coverage/language/nodejs/#npm
Your archive doesn't contain this folder.
That is why Trivy added this notification.

**Reply by eliassal** [reply]
Thanks, yes, it is working like a charm now But tell me, what is meant by "We exclude dev deps"? I work mainly with DotNet application, does this mean that when we try to scan source in dev state? In this case if we build for release then it will take into consideration the app for scanning. Also, I tried to scan a project (a little bit old) which has package-lock.json, when I ran trivy I got: (I am not sure if it an error related to npm or trivy is complaining that modules are not correctly versioned?) Thanks again

> ~/Projects/PartsUnlimitedWebsite$ trivy -d fs --include-dev-deps package-lock.json
> 2023-12-21T18:31:28.871+0100	DEBUG	Severities: ["UNKNOWN" "LOW" "MEDIUM" "HIGH" "CRITICAL"]
> 2023-12-21T18:31:28.874+0100	DEBUG	Ignore statuses	{"statuses": null}
> 2023-12-21T18:31:28.967+0100	DEBUG	cache dir:  /home/salam/.cache/trivy
> 2023-12-21T18:31:28.967+0100	DEBUG	DB update was skipped because the local DB is the latest
> 2023-12-21T18:31:28.967+0100	DEBUG	DB Schema: 2, UpdatedAt: 2023-12-21 12:12:32.605102755 +0000 UTC, NextUpdate: 2023-12-21 18:12:32.605102384 +0000 UTC, DownloadedAt: 2023-12-21 12:15:59.337571974 +0000 UTC
> 2023-12-21T18:31:28.967+0100	INFO	Vulnerability scanning is enabled
> 2023-12-21T18:31:28.967+0100	DEBUG	Vulnerability type:  [os library]
> 2023-12-21T18:31:28.967+0100	INFO	Secret scanning is enabled
> 2023-12-21T18:31:28.967+0100	INFO	If your scanning is slow, please try '--scanners vuln' to disable secret scanning
> 2023-12-21T18:31:28.967+0100	INFO	Please see also https://aquasecurity.github.io/trivy/v0.48/docs/scanner/secret/#recommendation for faster secret detection
> 2023-12-21T18:31:28.967+0100	DEBUG	Enabling misconfiguration scanners: [azure-arm cloudformation dockerfile helm kubernetes terraform terraformplan]
> 2023-12-21T18:31:28.967+0100	DEBUG	No secret config detected: trivy-secret.yaml
> 2023-12-21T18:31:28.967+0100	DEBUG	The nuget packages directory couldn't be found. License search disabled
> 2023-12-21T18:31:28.968+0100	DEBUG	Walk the file tree rooted at 'package-lock.json' in parallel
> 2023-12-21T18:31:28.968+0100	INFO	To collect the license information of packages in "package-lock.json", "npm install" needs to be performed beforehand
> 2023-12-21T18:31:29.072+0100	DEBUG	OS is not detected.
> 2023-12-21T18:31:29.072+0100	DEBUG	Detected OS: unknown
> 2023-12-21T18:31:29.072+0100	INFO	Number of language-specific files: 1
> 2023-12-21T18:31:29.072+0100	INFO	Detecting npm vulnerabilities...
> 2023-12-21T18:31:29.072+0100	DEBUG	Detecting library vulnerabilities, type: npm, path: package-lock.json
> 2023-12-21T18:31:29.072+0100	WARN	npm version error (https://registry.npmjs.org/ansi-regex/-/ansi-regex-0.1.0.tgz): invalid semantic version
> 2023-12-21T18:31:29.072+0100	WARN	npm version error (https://registry.npmjs.org/ansi-regex/-/ansi-regex-0.2.1.tgz): invalid semantic version
> 2023-12-21T18:31:29.073+0100	WARN	npm version error (https://registry.npmjs.org/ansi-regex/-/ansi-regex-1.1.1.tgz): invalid semantic version
> 2023-12-21T18:31:29.073+0100	WARN	npm version error (https://registry.npmjs.org/ansi-regex/-/ansi-regex-2.1.1.tgz): invalid semantic version
> 2023-12-21T18:31:29.073+0100	WARN	npm version error (https://registry.npmjs.org/async/-/async-0.1.22.tgz): invalid semantic version
> 2023-12-21T18:31:29.073+0100	WARN	npm version error (https://registry.npmjs.org/async/-/async-0.2.10.tgz): invalid semantic version
> 2023-12-21T18:31:29.073+0100	WARN	npm version error (https://registry.npmjs.org/async/-/async-0.9.2.tgz): invalid semantic version
> 2023-12-21T18:31:29.074+0100	WARN	npm version error (https://registry.npmjs.org/async/-/async-1.5.2.tgz): invalid semantic version
> 2023-12-21T18:31:29.074+0100	WARN	npm version error (https://registry.npmjs.org/bl/-/bl-0.9.5.tgz): invalid semantic version
> 2023-12-21T18:31:29.074+0100	WARN	npm version error (https://registry.npmjs.org/bl/-/bl-0.9.5.tgz): invalid semantic version
> 2023-12-21T18:31:29.074+0100	WARN	npm version error (https://registry.npmjs.org/bl/-/bl-0.9.5.tgz): invalid semantic version
> 2023-12-21T18:31:29.074+0100	WARN	npm version error (https://registry.npmjs.org/bower/-/bower-1.3.12.tgz): invalid semantic version
> 2023-12-21T18:31:29.074+0100	WARN	npm version error (https://registry.npmjs.org/bower/-/bower-1.3.12.tgz): invalid semantic version
>

**Reply by DmitriyLewen** [reply]
Typically developer dependencies are not included to resulting app (In some case lock files don't include dev deps).
e.g. for npm - https://docs.npmjs.com/specifying-dependencies-and-devdependencies-in-a-package-json-file
This way, dev deps have no influence on your app.

Thit is why we exclude dev deps from report by default (Users asked to include dev deps for npm and yarn, and we added `--include-dev-deps` flag for them).

> Also, I tried to scan a project (a little bit old) which has package-lock.json, when I ran trivy I got: (I am not sure if it an error related to npm or trivy is complaining that modules are not correctly versioned?) Thanks again

Can you share this `package-lock.json` file (or link to this project)?
It looks like `version` is using URL.
But usually `npm` sets the semver version in `version` field and moves URL to `resolved` field.

**Comment by eliassal** [comment]
Thanks for your help, I think this is the reason because I come from Dotnetworld and that is why I was confused. You can download the whole project at https://transfert.free.fr/DY2o1F6

**Comment by eliassal** [comment]
Or if you prefer, here is the package-lock.json only
[package-lock.json](https://github.com/aquasecurity/trivy/files/13750975/package-lock.json)

**Reply by DmitriyLewen** [reply]
I was right.
This `package-lock.json` file uses url as version.
But you can use [npm i --package-lock-only](https://docs.npmjs.com/cli/v10/commands/npm-install#package-lock-only) command to update lock file version and resolve versions:
```bash
root@bcf18acaf98e:/app# cat package-lock.json
{
  "name": "PartsUnlimited",
  "version": "0.0.0",
  "lockfileVersion": 1,
  "requires": true,
  "dependencies": {
    "abbrev": {
      "version": "https://registry.npmjs.org/abbrev/-/abbrev-1.1.1.tgz",
      "integrity": "sha1-+PLIh60Qv2f2NPAFtph/7TF5qsg="
    }
  }
}

root@bcf18acaf98e:/app# npm ci
npm WARN old lockfile 
npm WARN old lockfile The package-lock.json file was created with an old version of npm,
npm WARN old lockfile so supplemental metadata must be fetched from the registry.
npm WARN old lockfile 
npm WARN old lockfile This is a one-time fix-up, please be patient...
npm WARN old lockfile 

added 1 package, and audited 2 packages in 541ms

found 0 vulnerabilities
root@bcf18acaf98e:/app# npm i --package-lock-only
npm WARN old lockfile 
npm WARN old lockfile The package-lock.json file was created with an old version of npm,
npm WARN old lockfile so supplemental metadata must be fetched from the registry.
npm WARN old lockfile 
npm WARN old lockfile This is a one-time fix-up, please be patient...
npm WARN old lockfile 

up to date, audited 2 packages in 552ms

found 0 vulnerabilities
root@bcf18acaf98e:/app# cat package-lock.json 
{
  "name": "app",
  "version": "1.0.0",
  "lockfileVersion": 3,
  "requires": true,
  "packages": {
    "": {
      "name": "app",
      "version": "1.0.0",
      "license": "ISC",
      "dependencies": {
        "abbrev": "^1.1.1"
      },
      "devDependencies": {}
    },
    "node_modules/abbrev": {
      "version": "1.1.1",
      "resolved": "https://registry.npmjs.org/abbrev/-/abbrev-1.1.1.tgz",
      "integrity": "sha1-+PLIh60Qv2f2NPAFtph/7TF5qsg="
    }
  }
}
root@bcf18acaf98e:/app# 

```

**Reply by eliassal** [reply]
So many thanks I really appreciate your help, I learned a lot with your responses. Yes, after running the commands I was able to see the analysis in html format. One last question, how trivy compares to competitors like Checkmarks, Jfrog, Snyk....? Why I am asking this is because I worked for a public administration which had checkmarx but I understood it was very expensive.

**Reply by DmitriyLewen** [reply]
I only work with Trivy, unfortunately I can’t give a correct assessment
